File: ios/Sources/ACCore/Models/Elements/ProgressElements.swift
// Copyright (c) Microsoft Corporation. All rights reserved.
// Author: Vikrant Singh (github.com/VikrantSingh01)
// Licensed under the MIT License.

import Foundation
// MARK: - ProgressBar

public struct ProgressBar: Codable, Equatable {
    public let type: String = "ProgressBar"
    public var id: String?
    public var value: Double?
    public var max: Double?
    public var label: String?
    public var color: String?
    public var spacing: Spacing?
    public var separator: Bool?
    public var height: BlockElementHeight?
    public var isVisible: Bool?
    public var requires: [String: String]?

    /// Normalized progress fraction (0.0–1.0).
    /// - When `max` is explicitly set: `value / max`
    /// - When `max` is nil and `value` <= 1: treat value as already a 0–1 fraction
    /// - When `max` is nil and `value` > 1: treat as 0–100 scale
    public var normalizedValue: Double {
        let v = value ?? 0
        if let m = max {
            guard m > 0 else { return 0 }
            return Swift.min(Swift.max(v / m, 0), 1)
        }
        // No explicit max: auto-detect scale
        if v >= 0 && v <= 1 {
            return Swift.max(v, 0)
        }
        return Swift.min(Swift.max(v / 100, 0), 1)
    }

    public init(
        id: String? = nil,
        value: Double? = nil,
        max: Double? = nil,
        label: String? = nil,
        color: String? = nil,
        spacing: Spacing? = nil,
        separator: Bool? = nil,
        height: BlockElementHeight? = nil,
        isVisible: Bool? = nil,
        requires: [String: String]? = nil
    ) {
        self.id = id
        self.value = value
        self.max = max
        self.label = label
        self.color = color
        self.spacing = spacing
        self.separator = separator
        self.height = height
        self.isVisible = isVisible
        self.requires = requires
    }
}

// MARK: - ProgressRing

public struct ProgressRing: Codable, Equatable {
    public let type: String = "ProgressRing"
    public var id: String?
    public var label: String?
    public var labelPosition: String?
    public var size: String?
    public var color: String?
    public var horizontalAlignment: HorizontalAlignment?
    public var spacing: Spacing?
    public var separator: Bool?
    public var height: BlockElementHeight?
    public var isVisible: Bool?
    public var requires: [String: String]?

    public init(
        id: String? = nil,
        label: String? = nil,
        labelPosition: String? = nil,
        size: String? = nil,
        color: String? = nil,
        horizontalAlignment: HorizontalAlignment? = nil,
        spacing: Spacing? = nil,
        separator: Bool? = nil,
        height: BlockElementHeight? = nil,
        isVisible: Bool? = nil,
        requires: [String: String]? = nil
    ) {
        self.id = id
        self.label = label
        self.labelPosition = labelPosition
        self.size = size
        self.color = color
        self.horizontalAlignment = horizontalAlignment
        self.spacing = spacing
        self.separator = separator
        self.height = height
        self.isVisible = isVisible
        self.requires = requires
    }
}

// MARK: - Spinner

public struct Spinner: Codable, Equatable {
    public let type: String = "Spinner"
    public var id: String?
    public var size: SpinnerSize?
    public var label: String?
    public var spacing: Spacing?
    public var separator: Bool?
    public var height: BlockElementHeight?
    public var isVisible: Bool?
    public var requires: [String: String]?

    public init(
        id: String? = nil,
        size: SpinnerSize? = nil,
        label: String? = nil,
        spacing: Spacing? = nil,
        separator: Bool? = nil,
        height: BlockElementHeight? = nil,
        isVisible: Bool? = nil,
        requires: [String: String]? = nil
    ) {
        self.id = id
        self.size = size
        self.label = label
        self.spacing = spacing
        self.separator = separator
        self.height = height
        self.isVisible = isVisible
        self.requires = requires
    }
}

File: android/ac-core/src/main/kotlin/com/microsoft/adaptivecards/core/models/ProgressElements.kt
// Copyright (c) Microsoft Corporation. All rights reserved.
// Author: Vikrant Singh (github.com/VikrantSingh01)
// Licensed under the MIT License.

package com.microsoft.adaptivecards.core.models

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.Transient
import kotlinx.serialization.json.JsonElement

@Serializable
@SerialName("ProgressBar")
data class ProgressBar(
    @Transient override val type: String = "ProgressBar",
    override val id: String? = null,
    override val isVisible: Boolean = true,
    override val separator: Boolean = false,
    override val spacing: Spacing? = null,
    override val height: BlockElementHeight? = null,
    override val requires: Map<String, String>? = null,
    override val fallback: JsonElement? = null,
    val value: Double = 0.0,
    val max: Double? = null,
    val label: String? = null,
    val color: String? = null
) : CardElement {
    /** Normalized progress fraction (0.0–1.0).
     * When max is explicitly set: value / max.
     * When max is null and value <= 1: treat as 0–1 fraction.
     * When max is null and value > 1: treat as 0–100 scale. */
    val normalizedValue: Double
        get() {
            val m = max
            if (m != null) {
                return if (m > 0) (value / m).coerceIn(0.0, 1.0) else 0.0
            }
            return if (value in 0.0..1.0) {
                value.coerceAtLeast(0.0)
            } else {
                (value / 100.0).coerceIn(0.0, 1.0)
            }
        }
}

@Serializable
@SerialName("ProgressRing")
data class ProgressRing(
    @Transient override val type: String = "ProgressRing",
    override val id: String? = null,
    override val isVisible: Boolean = true,
    override val separator: Boolean = false,
    override val spacing: Spacing? = null,
    override val height: BlockElementHeight? = null,
    override val requires: Map<String, String>? = null,
    override val fallback: JsonElement? = null,
    val label: String? = null,
    val labelPosition: String? = null,
    val size: String? = null,
    val color: String? = null,
    val horizontalAlignment: HorizontalAlignment? = null
) : CardElement

@Serializable
@SerialName("Spinner")
data class Spinner(
    @Transient override val type: String = "Spinner",
    override val id: String? = null,
    override val isVisible: Boolean = true,
    override val separator: Boolean = false,
    override val spacing: Spacing? = null,
    override val height: BlockElementHeight? = null,
    override val requires: Map<String, String>? = null,
    override val fallback: JsonElement? = null,
    val size: SpinnerSize? = null,
    val label: String? = null
) : CardElement
